http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/ProxyWireMockTest.java

```diff
@@ -0,0 +1,137 @@
+ */
+
+
+package software.amazon.awssdk.http.nio.netty;
+
+import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
+import static com.github.tomakehurst.wiremock.client.WireMock.get;
+import static com.github.tomakehurst.wiremock.client.WireMock.urlPathEqualTo;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import com.github.tomakehurst.wiremock.WireMockServer;
+import com.github.tomakehurst.wiremock.core.WireMockConfiguration;
+import java.io.IOException;
+import java.util.concurrent.CompletionException;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpMethod;
+import software.amazon.awssdk.http.async.AsyncExecuteRequest;
+import software.amazon.awssdk.http.async.SdkAsyncHttpClient;
+import software.amazon.awssdk.http.async.SdkAsyncHttpResponseHandler;
+
+/**
+ * Tests for HTTP proxy functionality in the Netty client.
+ */
+public class ProxyWireMockTest {
+    private static SdkAsyncHttpClient client;
+
+    private static ProxyConfiguration proxyCfg;
+
+    private static WireMockServer mockServer = new WireMockServer(new WireMockConfiguration()
+            .dynamicPort()
+            .dynamicHttpsPort());
+
+    private static WireMockServer mockProxy = new WireMockServer(new WireMockConfiguration()
+            .dynamicPort()
+            .dynamicHttpsPort());
+
+    @BeforeClass
+    public static void setup() {
+        mockProxy.start();
+        mockServer.start();
+
+        mockServer.stubFor(get(urlPathEqualTo("/")).willReturn(aResponse().withStatus(200).withBody("hello")));
+
+        proxyCfg = ProxyConfiguration.builder()
+                .host("localhost")
+                .port(mockProxy.port())
+                .build();
+    }
+
+    @AfterClass
+    public static void teardown() {
+        mockServer.stop();
+        mockProxy.stop();
+    }
+
+    @After
+    public void methodTeardown() {
+        if (client != null) {
+            client.close();
+        }
+        client = null;
+    }
+
+    @Test(expected = IOException.class)
+    public void proxyConfigured_attemptsToConnect() throws Throwable {
+        AsyncExecuteRequest req = AsyncExecuteRequest.builder()
+                .request(testSdkRequest())
+                .responseHandler(mock(SdkAsyncHttpResponseHandler.class))
+                .build();
+
+        client = NettyNioAsyncHttpClient.builder()
+                .proxyConfiguration(proxyCfg)
+                .build();
+
+        try {
+            client.execute(req).join();
+        } catch (CompletionException e) {
+            Throwable cause = e.getCause();
+            // WireMock doesn't allow for mocking the CONNECT method so it will just return a 404, causing the client
```

Comment:
Seems like a big and unfortunate gap in our functional tests. Are there any other options to exercise the happy code path end to end?

Reply:
Yeah this is the first time we've needed to do this level of testing related to proxy. I'm not aware of a library that we can use for mocking this out. We'll either need to search for one, or build our own mini proxy server.